Jc/per project deps in properties

jmcampanini opened the pull request:
// for feedback: @uschi2000 + @qinfchen

it adds project specific syntax to the properties file

#### TODO
- [x] maybe improve the syntax for the project specific entry? it's `-projectName-` right now :(
- [x] add README - nothing in the README, would need to modify the wiki :(
- [ ] check if it works in their CI - for some reason build fails locally, but tests pass in IDE
- [ ] write an integration test for a multi-project setup

jmcampanini commented on src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy:
```diff
@@ -43,12 +63,53 @@ class PropertyFileRecommendationProviderSpec extends Specification {
+            some:other = $com.google.guava:guava
+            -notProjectName-GUAVA_VERSION = 11.0
+            -projectName-GUAVA_VERSION = 99.0
+            -projectName-com.google.guava:guava = $GUAVA_VERSION
```
it actually allows for per-project variables to be resolved.. nifty!

uschi2000 commented on src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java:
```diff
@@ -12,14 +12,20 @@
         protected String propertyValue(String name) {
-            return recommendations.getProperty(name);
+            String key = "-" + project.getName() + "-" + name;
+            String rv = recommendations.getProperty(key);
```
````
return Objects.firstNonNull(recommendations.getProperty(key), recommendations.getProperty(name));
````

jmcampanini replied:
the method can and should return `null` for it to function properly, and firstNoNull throws an NPE if both are null :(

uschi2000 replied:
oh, I see

On Tue, May 3, 2016 at 5:39 PM Javier Campanini <notifications@github.com>
wrote:

> In
> src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java
> <https://github.com/jmcampanini/nebula-dependency-recommender-plugin/pull/1#discussion_r61903405>
> :
>
> >          @Override
> >          protected Collection<String> propertyNames() {
> >              return recommendations.stringPropertyNames();
> >          }
> >
> >          @Override
> >          protected String propertyValue(String name) {
> > -            return recommendations.getProperty(
>
> name);
> > +            String key = "-" + project.getName() + "-" + name;
> > +            String rv = recommendations.getProperty(key);
>
> the method can and should return null for it to function properly, and
> firstNoNull throws an NPE if both are null :(
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jmcampanini/nebula-dependency-recommender-plugin/pull/1/files/3c94da02a20a07042d05f6c0f41ef8294c10c779#r61903405>
>

uschi2000 commented on src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java:
```diff
@@ -12,14 +12,20 @@
         @Override
         protected String propertyValue(String name) {
-            return recommendations.getProperty(name);
+            String key = "-" + project.getName() + "-" + name;
```
so the plugin already had support for the `-project-group:module` syntax, it just wasn't, uhm, "enabled"?

jmcampanini replied:
yea. they had all the pieces, just didn't put it together that way...

uschi2000 commented on src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy:
```diff
@@ -17,18 +17,38 @@ class PropertyFileRecommendationProviderSpec extends Specification {
+
+        propFile << '''
+            com.google.guava:guava = 18.0
+            -notProjectName-com.google.guava:guava = 99.0
```
should be != 99

uschi2000 commented on src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy:
```diff
@@ -17,18 +17,38 @@ class PropertyFileRecommendationProviderSpec extends Specification {
-        propFile << 'com.google.guava:guava = 18.0'
+        propFile << '''
+            com.google.guava:guava = 18.0
+            -notProjectName-com.google.guava:guava = 99.0
```
should be != 99

uschi2000 commented:
this is cool, let's propose it to them.

jmcampanini commented:
addressed your comments. going to modify the README and write an integration test for a multi-project setup, just to make sure it works. then will submit a PR to nebula

uschi2000 commented:
nice

On Tue, May 3, 2016 at 6:02 PM Javier Campanini <notifications@github.com>
wrote:

> addressed your comments. going to modify the README and write an
> integration test for a multi-project setup, just to make sure it works.
> then will submit a PR to nebula
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jmcampanini/nebula-dependency-recommender-plugin/pull/1#issuecomment-216577193>
>

qinfchen commented on src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java:
```diff
@@ -19,7 +21,12 @@
         @Override
         protected String propertyValue(String name) {
-            return recommendations.getProperty(name);
+            String projectSpecificName = project.getName() + PROJECT_DELIM + name;
```
what does this mean for root project? how does this play with global settings? is there any difference?
